### PR TITLE
feat: better type-safety for locales

### PR DESCRIPTION
**Pull request by mieszkosabo**
Hi, I updated the type of `createI18n` in a non-breaking way, so that it is possible to get better type-safety and IDE autocompletion on locales.

So for example given:
```ts
// locales/index.ts
import { createI18n } from 'next-international';
import type Locale from './en';

const locales = {
  en: () => import('./en'),
  fr: () => import('./fr'),
};

export const { useI18n, useScopedI18n, I18nProvider, useChangeLocale, defineLocale, getLocaleProps, useCurrentLocale } =
  createI18n<typeof Locale, typeof locales>(locales);
```

We get `useChangeLocale` to be typed as `() => (newLocale: "en" | "fr") => void` instead of `() => (newLocale: string) => void`

After I did that I thought that a typed wrapper for `useRouter`'s locale would be useful as well so I added simple `useCurrentLocale` which also has precise types instead of `string`.

I updated the README (sorry for added semicolons, my IDE picked up the prettier config, I can remove them if you'd like).

I also updated the example to use this second generic argument.

pnpm generated new lock file: I'm new to pnpm, not sure why this happened, I can remove this change.

Please let me know what you think!

### CODE REVIEW

File: packages/next-international/src/i18n/create-i18n.ts [hunk 1]
```diff
@@ -1,21 +1,23 @@
...
 import { createDefineLocale } from './create-define-locale';
 import { createGetLocaleProps } from './create-get-locale-static-props';
 import { createI18nProvider } from './create-i18n-provider';
 import { createUseChangeLocale } from './create-use-change-locale';
 import { createUsei18n } from './create-use-i18n';
 import { createScopedUsei18n } from './create-use-scoped-i18n';
+import { createUseCurrentLocale } from './create-use-current-locale';
 
-export function createI18n<Locale extends BaseLocale>(locales: Locales) {
+export function createI18n<Locale extends BaseLocale, Locales extends LocalesBase = LocalesBase>(locales: Locales) {
```
Comment (QuiiBz) [inline packages/next-international/src/i18n/create-i18n.ts:12]: I don't think we need this generic, as we can already have access to the type of the locales using the `locales` parameter

File: packages/next-international/README.md [hunk 2]
```diff
@@ -44,13 +44,21 @@ pnpm install next-international
 2. Create `locales/index.ts` with your locales:
 
 ```ts
-import { createI18n } from 'next-international'
-import type Locale from './en'
+import { createI18n } from 'next-international';
```
Comment (QuiiBz) [inline packages/next-international/README.md:47]: Could you remove each added semicolon?

**Review by QuiiBz — CHANGES_REQUESTED** [review]
Great idea, thanks! I've left some feedback, could you also revert the `pnpm-lock.yaml` changes?

**Comment by mieszkosabo** [comment]
Hey @QuiiBz, so as I said, all generic parameters need to be either inferred or specified by the user. But that got me thinking why don't we infer everything then. So I added a commit which does just that. I think it's clean, because the user gets type safety on locales and they don't need to provide the redundant type arguments to `createI18n`.

This is however a breaking change now. Luckily migration is pretty trivial, users just need to remove the generic argument:
```ts
import Locale from "./en";

createI18n<typeof Locale>({
    en: () => import('./en'),
    fr: () => import('./fr'),
  });
```
to
```ts
createI18n({
    en: () => import('./en'),
    fr: () => import('./fr'),
  });
```

Also explicit typing of eg. JSON files needs a slightly different syntax:
```ts
import { createI18n } from 'next-international';

type Locale = {
  hello: string;
  welcome: string;
}

type Locales = {
  en: Locale;
  fr: Locale;
}

export const {
  ...
} = createI18n<any, Locales>({
  en: () => import('./en.json'),
  fr: () => import('./fr.json'),
})
```

File: packages/international-types/index.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 export type LocaleValue = string | number | boolean | null | undefined | Date;
 export type BaseLocale = Record<string, LocaleValue>;
+export type BaseLocales = Record<string, any>;
```
Comment (QuiiBz) [inline packages/international-types/index.ts:3]: Could we restrict more this type? We only want to accept objects with `() => import(...)` statements:

```suggestion
export type BaseLocales = Record<string, () => Promise<{ default: any }>>;
```

File: packages/international-types/index.ts [hunk 4]
```diff
@@ -47,3 +48,22 @@ export type ScopedValue<
...
+
+// Given a object type with string keys, return the "first" key.
+// Because the key ordering is not guaranteed, this type should be used
+// only when the key order is not important.
+type SomeKey<T extends Record<string, any>> = UnionToTuple<keyof T>[0] extends string
+  ? UnionToTuple<keyof T>[0]
+  : never;
+
+// Gets a single locale type from an object of the shape of BaseLocales.
+export type GetLocaleType<Locales extends BaseLocales> = Locales[SomeKey<Locales>] extends () => any
```
Comment (QuiiBz) [inline packages/international-types/index.ts:67]: Same here:

```suggestion
export type GetLocaleType<Locales extends BaseLocales> = Locales[SomeKey<Locales>] extends () => Promise<{ default: any }>
```

**Review by QuiiBz — CHANGES_REQUESTED** [review]

**Comment by QuiiBz** [comment]
@mieszkosabo the `GetLocaleType` type is very smart, I've never thought of doing it this way. It's a lot better to not have to provide the generic manually!

As this is a breaking change, I'll create a new minor release (since we're not yet in 1.0) after merging this PR.

**Comment by mieszkosabo** [comment]
Thanks for the feedback! I refactored the types a little bit – I think it's better now. I added tests for the types to check if it works for both cases (imported types and hand written types) and I noticed some bugs, but hopefully they're all fixed now 👍🏻.

**Review by QuiiBz — APPROVED** [review]
LGTM, thanks a lot for your work!

**Pull request merged**